packages/monitoring-system/data-processor/src/data-processors/github-data-processor.ts

```diff
@@ -12,10 +12,169 @@
+}
+
+/**
+ * Collects and processes Events data from GitHub
+ */
 export class GitHubProcessor extends DataProcessor {
+  private octokit: Octokit;
+
+  constructor(options?: GitHubProcessorOptions) {
+    super(options);
+    this.octokit = options?.octokit || new Octokit();
+  }
+
+  /**
+   * Collect and process GitHub Events data
+   */
   public async collectAndProcess(): Promise<void> {
-    throw new Error('Method not implemented.');
+    return new Promise((resolve, reject) => {
+      this.listRepositories()
+        .then(repos => {
+          return Promise.all(
+            repos.map(repo => {
+              return this.listPublicEventsForRepository(repo);
+            })
+          );
+        })
+        .then((allRepoEvents: GitHubEventDocument[][]) => {
+          const allRepoEventsFlattened = ([] as GitHubEventDocument[]).concat(
+            ...allRepoEvents
+          );
+          return this.storeEventsData(allRepoEventsFlattened);
+        })
+        .then(() => resolve())
+        .catch(error => {
+          console.error(`Failed to process GitHub Events data: ${error}`);
+          reject(new Error(`Failed to process GitHub Events data: ${error}`));
+        });
+    });
+  }
+
+  /**
+   * List the GitHub repositories that have triggered
+   * bot executions in the past
+   */
+  private async listRepositories(): Promise<GitHubRepositoryDocument[]> {
+    return this.firestore
+      .collection('GitHub_Repository')
+      .get()
+      .then(repositoryCollection => {
+        const repositoryDocs = repositoryCollection.docs;
+        return repositoryDocs.map(
+          repoDoc => repoDoc.data() as GitHubRepositoryDocument
+        );
+      });
+  }
+
+  /**
+   * Get all the publicly visible Events on the given repository
+   * @param repository repository for which to get events
+   */
+  private async listPublicEventsForRepository(
+    repository: GitHubRepositoryDocument
+  ): Promise<GitHubEventDocument[]> {
+    return this.octokit.activity
+      .listRepoEvents({
+        repo: repository.repo_name,
+        owner: repository.owner_name,
+      })
+      .then(eventsPayload => {
+        if (!(eventsPayload.data instanceof Array)) {
+          throw new Error(
+            `Unexpected payload from Octokit: ${JSON.stringify(eventsPayload)}`
+          );
+        }
+        const gitHubEvents: GitHubEventDocument[] = [];
+        for (const event of eventsPayload.data) {
+          gitHubEvents.push(
+            this.githubEventResponseToEvent(
+              (event as unknown) as GitHubEventResponse
+            )
+          );
+        }
+        return gitHubEvents;
+      });
+  }
+
+  /**
+   * Converts GitHub's list event response to a GitHubEventDocument
+   * @param eventResponse list event response from GitHub
+   */
+  private githubEventResponseToEvent(
+    eventResponse: GitHubEventResponse
+  ): GitHubEventDocument {
+    const {type, repo, payload, created_at, org, user} = eventResponse;
```

Comment:
I think this destructuring makes the code later harder to read. Some of these are only used in one place and later it would read easier to see that they are coming from fields on eventResponse

Reply:
got it, I can change that